### ISSUE TITLE
PlayerDrawLayer.Position Abstraction

### ISSUE THREAD

**Issue by absoluteAquarian** [bluemagic123/tModLoader#2752]
### Do you intend to personally contribute/program this feature?

No

### I would like to see this change made to improve my experience with

Mod capability as a Modder

### Description

A while ago, `PlayerDrawLayer` was rewritten so that layers are ordered via a dependency/dependent system instead of modifying a collection of layers.

I believe that a more generic implementation (in a literal sense) would be useful for mod developers who wish to implement such a sorting system without having copy/paste large swaths of code from tModLoader.

### What does this proposal attempt to solve or improve?

This proposal would allow mod developers to easily implement the PlayerDrawLayer sorting system on their own types.

For example, in Magic Storage, the backend is currently being rewritten to allow mod developers to add their own sorting and filtering options in the mod's Storage and Crafting UIs.  While I could simply use a `List<T>` and call it a day, this wouldn't be safe (mod developers should *not* be able to remove entries in the list).

### Which (other) solutions should be considered?

n/a

**Comment by Chicken-Bones** [comment]
Absolutely, we want to do something like this for other position dependent systems (like WorldGen passes).
I was previously unsure if the parenting system would be applicable to other systems, but it seems pretty solid. Just keep in mind the system is more than just sorting a flat list of layers.

Key features of the system to keep in mind when considering applicability to your own use case:
- Entries have default visibility.
- Entries can be hidden by mods, but not shown.
- A hidden entry also hides its children
- The `Multiple` ordering is used to position a layer in different places depending on conditions (also 'bringing' child layers with it)
- You can always make another layer in a different position, which calls to the same function, and hide the original layer. This won't move child layers (which may not wish to be moved)
- Entries cannot be moved
  - The `ModifyDrawLayerOrdering` hook does allow for it but with documentation "Use with extreme caution, or risk breaking other mods."
  - The primary use of this would be to adjust the conditions for the `Multiple` condition, but even then, it's unclear how to do this in a compatible way, and converting a layer position from `Between` to `Multiple` means that other layers which sort `Between` the changed layer will fail to sort correctly.
  
  
 Honestly, some feedback on this system is desirable. The question of how to handle layer moving is a challenging one. 
 - The child system does allow for layers to be moved, and bring their children with them. I think it's quite good.
 - Should we make it easier to 'dupe' layers conditionally? Or to convert a layer position from `Between` to `Multiple`, or add another condition to `Multiple`?
